include/mfmg/dealii/lanczos.templates.hpp

```diff
@@ -0,0 +1,384 @@
+  }
+
+  assert(_op.m() == _op.n() && "Operator must be square");
+
+  assert(_maxit >= 0);
+  assert(_tol >= 0.);
+}
+
+/// \brief Lanczos solver: accessor for (approximate) eigenvalue
+template <typename OperatorType, typename VectorType>
+std::vector<double> const &Lanczos<OperatorType, VectorType>::get_evals() const
+{
+  return _evals;
+}
+
+template <typename OperatorType, typename VectorType>
+double Lanczos<OperatorType, VectorType>::get_eval(int i) const
+{
+  assert(i >= 0 && (size_t)i < _evals.size());
+  return _evals[i];
+}
+
+/// \brief Lanczos solver: accessor for (approximate) eigenvector
+template <typename OperatorType, typename VectorType>
+VectorType const &Lanczos<OperatorType, VectorType>::get_evec(int i) const
+{
+  assert(i >= 0 && (size_t)i < _evecs.size());
+  return _evecs[i];
+}
+
+/// \brief Lanczos solver: accessor for (approximate) eigenvectors
+template <typename OperatorType, typename VectorType>
+std::vector<VectorType> const &
+Lanczos<OperatorType, VectorType>::get_evecs() const
+{
+  return _evecs;
+}
+
+/// \brief Lanczos solver: perform Lanczos solve, use random initial guess
+template <typename OperatorType, typename VectorType>
+void Lanczos<OperatorType, VectorType>::solve()
+{
+  // By default set initial guess to a random vector
+  VectorType initial_guess(_op.n());
+
+  if (!_is_deflated)
+  {
+    details_set_initial_guess(initial_guess);
+    details_solve_lanczos(_op, _num_requested, initial_guess, _evals, _evecs);
+  }
+  else
+  {
+    // Form deflated operator from original operator
+    DeflatedOperator<OperatorType, VectorType> deflated_op(_op);
+
+    // Loop over Lanczos solves
+    for (int cycle = 0; cycle < _num_cycles; ++cycle)
+    {
+      details_set_initial_guess(initial_guess, cycle);
+
+      deflated_op.deflate(initial_guess);
+
+      std::vector<double> evals;
+      std::vector<VectorType> evecs;
+      details_solve_lanczos(deflated_op, _num_evecs_per_cycle, initial_guess,
+                            evals, evecs);
+
+      // Save the eigenpairs just calculated
+
+      // NOTE: throughout we use the term eigenpair (= eigenvector,
+      // eigenvalue), though the precise terminology should be "approximate
+      // eigenpairs" or "Ritz pairs."
+      for (int i = 0; i < _num_evecs_per_cycle; ++i)
+      {
+        _evals.push_back(evals[i]);
+        _evecs.push_back(evecs[i]);
+      }
+
+      // Add eigenvectors to the set of vectors being deflated out
+      if (cycle != _num_cycles - 1)
+        deflated_op.add_deflation_vecs(evecs);
+    }
+  }
+}
+
+/// \brief Lanczos solver: perform Lanczos solve
+template <typename OperatorType, typename VectorType>
+template <typename FullOperatorType>
+void Lanczos<OperatorType, VectorType>::details_solve_lanczos(
+    FullOperatorType const &op, const int num_requested,
+    VectorType const &initial_guess, std::vector<double> &evals,
+    std::vector<VectorType> &evecs) const
```

Comment:
Readability: consider returning `evals` and `evecs` instead of taking them by reference.